### DISCUSSION TITLE
cannot import service module

### DISCUSSION THREAD

**Discussion by mvgarcia07** [marcelotduarte/cx_Freeze#892] [Q&A] [answered]
My application containing several modules with several inits on it and file dependency. 

I am having an issue is not finding my own modules, how should I tell in setup.py to include them?

Here are my logs, By the way, the logs are very helpful! 

```
[19684] 2021/01/20 19:52:49.446 starting logging at level ERROR
[19684] 2021/01/20 19:52:52.181 Python exception encountered:
[19684] 2021/01/20 19:52:52.181     Internal Message: cannot import service module
[19684] 2021/01/20 19:52:52.181     Type => <class 'ModuleNotFoundError'>
[19684] 2021/01/20 19:52:52.181     Value => No module named 'app.api.functions.func_gral'
[19684] 2021/01/20 19:52:52.183     Traceback (most recent call last):

[19684] 2021/01/20 19:52:52.183       File "<frozen zipimport>", line 259, in load_module

[19684] 2021/01/20 19:52:52.183       File "mypath\Model\ServiceHandler.py", line 12, in <module>
    from init import app

[19684] 2021/01/20 19:52:52.183       File "<frozen zipimport>", line 259, in load_module

[19684] 2021/01/20 19:52:52.183       File "mypath\Model\init.py", line 4, in <module>
    app = create_app('default')

[19684] 2021/01/20 19:52:52.183       File "mypathP\Model\app\__init__.py", line 13, in create_app
    from .api import api as api_blueprint

[19684] 2021/01/20 19:52:52.183       File "mypath\Model\app\api\__init__.py", line 5, in <module>
    from . import model

[19684] 2021/01/20 19:52:52.183       File "C:mypath\Model\app\api\model.py", line 3, in <module>
    from .functions.func_gral import *

[19684] 2021/01/20 19:52:52.183     ModuleNotFoundError: No module named 'app.api.functions.func_gral'

[19684] 2021/01/20 19:52:52.189 ending logging

```

my setup.py

```
from cx_Freeze import setup, Executable
import sys, os

options = {
    "build_exe": {
        "includes": ["ServiceHandler","cx_Logging"],
        "excludes": ["tkinter"], 
        "path": sys.path,
        "include_files": ['.\\app\\Models\\instanties\\sklear_randomforestclassifierv1.pkl', '.\\app\\api\\functions']
        
    }
}

```

**Comment by marcelotduarte** [comment, marked as answer]
Use "packages": ["app.api"] or similar, and remove the '.\\app\\api\\functions' in include_files.

**Reply by marcelotduarte** [reply]
https://cx-freeze.readthedocs.io/en/latest/distutils.html#build-exe

**Reply by mvgarcia07** [reply]
I did change the Setup as "app.api" and add the full path to the module  "app.api.functions.func_gral" but still not finding it even though is the exact same name. 

Any hints? 

```
[15956] 2021/01/20 22:05:13.153       File "`mypath\Model\app\api\model.py", line 3, in <module>
    from app.api.functions.func_gral import *

[15956] 2021/01/20 22:05:13.153     ModuleNotFoundError: No module named 'app.api.functions.func_gral'
``

**Reply by marcelotduarte** [reply]
```
options = {
    "build_exe": {
        "includes": ["ServiceHandler","cx_Logging"],
        "excludes": ["tkinter"], 
        "packages": ["app"],
        "path": sys.path,
        "include_files": ['.\\app\\Models\\instanties\\sklear_randomforestclassifierv1.pkl',]
        
    }
}
```

**Reply by marcelotduarte** [reply]
Maybe in "includes": ["ServiceHandler","cx_Logging", "app.api.functions.func_gral"],
There are a module with this name?

**Reply by mvgarcia07** [reply]
yes, I added the complete path so when compile it can find it, but is not working with any of these options.

![image](https://user-images.githubusercontent.com/16159114/105249455-09929c00-5b70-11eb-841c-ec17eb7d54ce.png)


![image](https://user-images.githubusercontent.com/16159114/105249292-cc2e0e80-5b6f-11eb-9f57-7daf5aa70f8c.png)

**Reply by marcelotduarte** [reply]
What is the Executable() line?

**Reply by mvgarcia07** [reply]
Sorry didn't get the question? the service Handler?

**Reply by marcelotduarte** [reply]
It is like this? 
executables = [ Executable("Config.py", base="Win32Service", target_name="FlaskService.exe") ]
And setup has it?

**Reply by mvgarcia07** [reply]
I have to renamed because I already have a config for my API, is that an issue?
`
executables = [
    Executable(
        "Configws.py",
        base="Win32Service",
        target_name="FlaskService.exe"
    )
]`

**Reply by marcelotduarte** [reply]
The name of the base config can be anyone. n the base console, the init script starts the import of the packages, and cx_Freeze can find the packages and modules. In the service the config has only variables, then the finder of packages and modules starts in the service handler.
Maybe a relative import in this ServiceHandler can be the cause of the trouble.

**Reply by mvgarcia07** [reply]
my init.py
```
from flask import Flask
from app import create_app

app = create_app('default')
```
my ServiceHanlder.py notice I was trying to add the system path but didn't work either.

```
import threading
import os
import sys
import cx_Logging
from init import app

#sys.path.append(os.path.dirname(__name__))
#os.chdir(sys._MEIPASS) #change to the current directory

class Handler:

    # no parameters are permitted; all configuration should be placed in the
    # configuration file and handled in the Initialize() method
    def __init__(self):
        self.stopEvent = threading.Event()
        self.stopRequestedEvent = threading.Event()

    # called when the service is starting
    def initialize(self, configFileName):
        self.directory = os.path.dirname(sys.executable)
        cx_Logging.StartLogging(os.path.join(self.directory, "teste.log"), cx_Logging.DEBUG)
        #pass

    # called when the service is starting immediately after Initialize()
    # use this to perform the work of the service; don't forget to set or check
    # for the stop event or the service GUI will not respond to requests to
    # stop the service
    def run(self):
        cx_Logging.Debug("stdout=%r", sys.stdout)
        sys.stdout = open(os.path.join(self.directory, "stdout.log"), "a")
        sys.stderr = open(os.path.join(self.directory, "stderr.log"), "a")
        app.run(host="127.0.0.1", port=5120)
        self.stopRequestedEvent.wait()
        self.stopEvent.set()

    # called when the service is being stopped by the service manager GUI
    def stop(self):
        self.stopRequestedEvent.set()
        self.stopEvent.wait()
```

**Reply by mvgarcia07** [reply]
I was able to overpass the module issue by adding` __init__.py ` in the functions directory. So I was able to compile and install.

**Comment by mvgarcia07** [comment]
@marcelotduarte I was able to overpass the module issue by adding `__init__.py` in the functions directory. So I was able to compile and install. 

However, the application stays in starting mode and is not running, so I can not see the logs.

Any hints on that one? Is any way I test the .exe without installing just to make sure is working?

Thanks in advance! 

Should I open a new Q&A?

![image](https://user-images.githubusercontent.com/16159114/105403925-48345f00-5c21-11eb-842e-53c7d1df6c58.png)

**Reply by marcelotduarte** [reply]
There are 4 logs (FlaskService.log, teste.log, stdout.log, stderr.log). Do you check all?
Services are hard to debug. You have to use print() and check stdout.log to see if it pass in the marked point.
But before you can test it as a console app, and move ahead to the service app.

**Reply by mvgarcia07** [reply]
For some reason, only FlagService.log is in there the other ones disappear.

I did create the console app and test it and it is working just fine except the logs are not showing. 

Then I created the service again and still stuck in starting...

Do you think could be something related to the flask making the service very slow? I am having the same behavior with pyinstaller but is throwing a timeout error.

I am running out of options here.

Thanks again for the support. 

```
from cx_Freeze import setup, Executable
import sys, os

options = {
    "build_exe": {
        "includes": ["cx_Logging","jinja2.ext"],
        "excludes": ["tkinter"],
         "packages": ["app"],
         "path": sys.path,
        "include_files": ['.\\app\\Models\\instanties\\sklear_randomforestclassifierv1.pkl']
        
    }
}

executables = [
    Executable(
        "ServiceHandler.py",
        base="Console",
        target_name="FlaskServiceConsole.exe"
    )
]
....

```

Here is my service handler

```

import threading
import os
import sys
import cx_Logging
#import waitress

from init import app

sys.path.append(os.path.dirname(__name__))
#os.chdir(sys._MEIPASS) #change to the current directory

class Handler:

    # no parameters are permitted; all configuration should be placed in the
    # configuration file and handled in the Initialize() method
    def __init__(self):
        self.stopEvent = threading.Event()
        self.stopRequestedEvent = threading.Event()

    # called when the service is starting
    def initialize(self, configFileName):
        self.directory = os.path.dirname(sys.executable)
        cx_Logging.StartLogging(os.path.join(self.directory, "teste.log"), cx_Logging.DEBUG)
        #pass

    # called when the service is starting immediately after Initialize()
    # use this to perform the work of the service; don't forget to set or check
    # for the stop event or the service GUI will not respond to requests to
    # stop the service
    def run(self):
        cx_Logging.Debug("stdout=%r", sys.stdout)
        sys.stdout = open(os.path.join(self.directory, "stdout.log"), "a")
        sys.stderr = open(os.path.join(self.directory, "stderr.log"), "a")
        app.run(host="127.0.0.1", port=5120)
        self.stopRequestedEvent.wait()
        self.stopEvent.set()

    # called when the service is being stopped by the service manager GUI
    def stop(self):
        self.stopRequestedEvent.set()
        self.stopEvent.wait()

if __name__ == "__main__":
    #PORT = sys.argv[1] if len(sys.argv) > 1 else 5120
    #waitress.serve(app, port=PORT)
    app.run(host="127.0.0.1", port=5120)
```

**Reply by mvgarcia07** [reply]
I was able to reproduce the error in simple template API, There is something wrong in the service handler that is not writing the logs when you have an API with submodules. Or probably the structure of my API??? 

Any suggestion I kindly appreciated, not sure if I should open this discussion regarding losing the logs and the service starting but not running. 

I already spent 5 days just getting this to work, I can not give up now. 

Here is my test
https://github.com/mvgarcia07/python-flask-as-window-service/tree/main/cx_Freeze_example_modules

**Reply by mvgarcia07** [reply]
@marcelotduarte  I was trying to fix the logs by giving a full path, after I did that, I was having an error regarding **numpy**. I think I still have the issue of not finding the modules.

I literally changed the logs and then I got the error. 

My only change in the ServiceHandler.py
```
#sys.stdout = open(os.path.join(self.directory, "stdout.log"), "a")
sys.stdout = open(os.path.join("C:\\Data\\logs\\", "stdout.log"), "a")
```
The error is 
```
Please note and check the following:

  * The Python version is: Python3.8 from "C:\cx_Freeze_example_modules\build\exe.win-amd64-3.8\FlaskService_modules.exe"
  * The NumPy version is: "1.19.5"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: DLL load failed while importing _multiarray_umath: The specified module could not be found.

```

I honestly need real-time support. Is anybody in your team available? how cab I contact them?

**Reply by marcelotduarte** [reply]
use numpy==1.18.2

**Reply by mvgarcia07** [reply]
The numpy error is gone thanks,

 But the logs are not in there and the services are hanging as "starting"


  
```
import threading
import os
import sys
import cx_Logging

from init import app


class Handler:

    # no parameters are permitted; all configuration should be placed in the
    # configuration file and handled in the Initialize() method
    def __init__(self):
        self.stopEvent = threading.Event()
        self.stopRequestedEvent = threading.Event()

    # called when the service is starting
    def initialize(self, configFileName):
        self.directory = os.path.dirname(sys.executable)
        #cx_Logging.StartLogging(os.path.join(self.directory, "teste.log"), cx_Logging.DEBUG)
        cx_Logging.StartLogging(os.path.join("C:\\Data\\logs\\", "teste.log"), cx_Logging.DEBUG)
        #pass

    # called when the service is starting immediately after Initialize()
    # use this to perform the work of the service; don't forget to set or check
    # for the stop event or the service GUI will not respond to requests to
    # stop the service
    def run(self):
        cx_Logging.Debug("stdout=%r", sys.stdout)
        #sys.stdout = open(os.path.join(self.directory, "stdout.log"), "a")
        #sys.stderr = open(os.path.join(self.directory, "stderr.log"), "a")
        sys.stdout = open(os.path.join("C:\\Data\\logs\\", "stdout.log"), "a")
        sys.stderr = open(os.path.join("C:\\Data\\logs\\", "stderr.log"), "a")
        app.run(host="127.0.0.1",port=5123)
        self.stopRequestedEvent.wait()
        self.stopEvent.set()

    # called when the service is being stopped by the service manager GUI
    def stop(self):
        self.stopRequestedEvent.set()
        self.stopEvent.wait()
```

**Reply by marcelotduarte** [reply]
Test as a console app before the continue with the service handler. A console app is easier to debug.

**Reply by mvgarcia07** [reply]
The console APP works but I can not see the logs. 
https://github.com/mvgarcia07/python-flask-as-window-service/tree/main/cx_Freeze_example_modules

```
from cx_Freeze import setup, Executable
import sys


options = {
    "build_exe": {
        "includes": ["cx_Logging","jinja2.ext"],
        "excludes": ["tkinter"],
         "packages": ["app"],
         "path": sys.path,
    }
}

executables = [
    Executable(
        "ServiceHandler_console.py",
        base="Console",
        target_name="FlaskService_modules_console.exe",
    )
]
.
.
.
```






´´´
import threading
import os
import sys
import cx_Logging
from init import app


class Handler:

    # no parameters are permitted; all configuration should be placed in the
    # configuration file and handled in the Initialize() method
    def __init__(self):
        self.stopEvent = threading.Event()
        self.stopRequestedEvent = threading.Event()

    # called when the service is starting
    def initialize(self, configFileName):
        self.directory = os.path.dirname(sys.executable)
        cx_Logging.StartLogging(os.path.join(self.directory, "teste.log"), cx_Logging.DEBUG)
        #cx_Logging.StartLogging(os.path.join("C:\\Data\\logs\\", "teste.log"), cx_Logging.DEBUG)
        #pass

    # called when the service is starting immediately after Initialize()
    # use this to perform the work of the service; don't forget to set or check
    # for the stop event or the service GUI will not respond to requests to
    # stop the service
    def run(self):
        cx_Logging.Debug("stdout=%r", sys.stdout)
        sys.stdout = open(os.path.join(self.directory, "stdout.log"), "a")
        sys.stderr = open(os.path.join(self.directory, "stderr.log"), "a")
        #sys.stdout = open(os.path.join("C:\\Data\\logs\\", "stdout.log"), "a")
        #sys.stderr = open(os.path.join("C:\\Data\\logs\\", "stderr.log"), "a")
        app.run(host="127.0.0.1",port=5123)
        self.stopRequestedEvent.wait()
        self.stopEvent.set()

    # called when the service is being stopped by the service manager GUI
    def stop(self):
        self.stopRequestedEvent.set()
        self.stopEvent.wait()
        
if __name__ == "__main__":
    app.run(host="127.0.0.1",port=5123)

´´´

**Reply by marcelotduarte** [reply]
I think that you are confused ;-)
You cannot mixup the console and the service. You can have the two in separate executables, in your setup:
```
executables = [
    Executable(
        "consoleapp.py",
        base="Console",
        target_name="FlaskConsole"
    ),
    Executable(
        "Config.py", # or other name, but not the ServiceHandler that has the class Handler
        base="Win32Service",
        target_name="FlaskService"
    )
]
```
#consoleapp.py
```
from init import app
if __name__ == "__main__":
    app.run(host="127.0.0.1",port=5123)
```

**Reply by mvgarcia07** [reply]
@marcelotduarte thanks for taking the time and the hint to compile both at the same time!, I fixed and it is working, I added the logs code to the **consoleapp.py** in my case the **init.py** in my case and I can see it, not a problem with that. But as I said there are no issues with the console. I know the console is working.

My problem is the window services not starting and not having logs to troubleshoot.  
https://github.com/mvgarcia07/python-flask-as-window-service/blob/main/cx_Freeze_example_modules/setup.py
https://github.com/mvgarcia07/python-flask-as-window-service/blob/main/cx_Freeze_example_modules/ServiceHandler.py

 I am in desperate mode. I feel I am close, Probably I am doing something stupid and I don't see it because as you can tell I am not an expert in windows services or creating .exe from python.

I am trying to understand why it is working with a simple flask and not with a flask app with submodules and dependencies that is a well as simple.

Any hint I kindly appreciated.  I need to conclude if this is possible with cx_freeze by tomorrow.

**Reply by marcelotduarte** [reply]
I would love to help you more, but I don't know if I can, but I'll try, that we can't give up...
Ah, upgrade to cx_Freeze 6.5.3
It is probably a problem with some package or something not allowed in services. For example, in console mode, sys.stdout and sys.stderr are shown on the screen, but in service mode they are None, so we had to create a file to save them.
And if some package check for input in sys.stdin, the service breaks.
Another thing can be a socket timeout. You can try use socket.setdefaulttimeout(10) in the initialize method.

There are a possibility to work around some of these issues using multiprocessing, and that console app will be useful.
Are you familiar with multiprocessing? Take a look in docs before read the following thread:
https://github.com/marcelotduarte/cx_Freeze/issues/536
There are 2 cases of success in that.

**Reply by mvgarcia07** [reply]
Did the upgrade,  and changed `AUTO_START = False` and added` socket.setdefaulttimeout(10)` 
I start the service manually and  I was able to see the timeout error;

**Error 1053: The Service Did Not Respond to the Start...**   This is the same error I was having with pyinstaller.

I understand the Error 1053 is because the application is taking more than 30 sec to launch. What is strange is that keeps hanging there and no stopping, I have to restart the machine to delete the application because `--uninstall test` is not enough.

I am not familiar with multiprocessing, I can give it a try if, you think I reached a dead-end having the service handler as it is.

**Reply by mvgarcia07** [reply]
@marcelotduarte thanks for the support.
I think I am done, if you have the time or the team here is the code. I look all over the internet, not body seems to have a solution neither for cx_freeze or pyinstaller. 

I narrowed it down to a simple example so is easy to troubleshoot. 
https://github.com/mvgarcia07/python-flask-as-window-service/tree/main/cx_Freeze_example_modules
